src/CMakeLists.txt

```
@@ -31,7 +31,7 @@ target_include_directories(
 set(JSON_MultipleHeaders
     OFF
     CACHE INTERNAL "")
-target_link_libraries(${PROJECT_NAME} PUBLIC MQT::qfr_dd)
+target_link_libraries(${PROJECT_NAME} PUBLIC MQT::Core MQT::CoreDD)
```

```suggestion
target_link_libraries(${PROJECT_NAME} PUBLIC MQT::CoreDD)
```
The specific targets already subsume the Core target.

---

Then there seems to be a flaw. Compiling with the proposed change produces the following error

```
[1/33] Building CXX object src/CMakeFiles/ddsim.dir/Simulator.cpp.o
FAILED: src/CMakeFiles/ddsim.dir/Simulator.cpp.o 
/usr/bin/g++-10  -I/home/stefan/repos/quantum/ddsim-public/include -I/home/stefan/repos/quantum/ddsim-public/extern/mqt-core/include/dd -I/home/stefan/repos/quantum/ddsim-public/cmake-build-debug/src/extern/mqt-core/include/dd -I/home/stefan/repos/quantum/ddsim-public/extern/mqt-core/include -I/home/stefan/repos/quantum/ddsim-public/cmake-build-debug/src/extern/mqt-core/include -isystem /home/stefan/repos/quantum/ddsim-public/extern/mqt-core/extern/json/single_include -isystem /home/stefan/repos/quantum/ddsim-public/extern/taskflow -g -flto=auto -fno-fat-lto-objects -fPIC -fdiagnostics-color=always -g -mtune=native -march=native -fno-omit-frame-pointer -fno-optimize-sibling-calls -fno-inline-functions -fvisibility=hidden -Wall -Wextra -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wno-unknown-pragmas -Wno-pragmas -Wmisleading-indentation -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wuseless-cast -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=noreturn -Wmissing-noreturn -Wsuggest-attribute=malloc -std=gnu++17 -MD -MT src/CMakeFiles/ddsim.dir/Simulator.cpp.o -MF src/CMakeFiles/ddsim.dir/Simulator.cpp.o.d -o src/CMakeFiles/ddsim.dir/Simulator.cpp.o -c /home/stefan/repos/quantum/ddsim-public/src/Simulator.cpp
In file included from /home/stefan/repos/quantum/ddsim-public/src/Simulator.cpp:1:
/home/stefan/repos/quantum/ddsim-public/include/Simulator.hpp:62:41: error: ‘Qubit’ in namespace ‘qc’ does not name a type
   62 |     char measureOneCollapsing(const qc::Qubit index, const bool assumeProbabilityNormalization = true) {
      |                                         ^~~~~
```

---

If I had a guess, then this might be due to an error in some include.
Due to just copying files as part of the transition towards MQT Core, there are now three files called `Definition.hpp` (one in the ZX and the DD sub package and one in the top-level package).
Due to the way includes are resolved, it could be that one of the "#include "Definitions.hpp" calls is resolved to "#include "dd/Definitions.hpp" instead of the general definitions header.

I plan to work around this awkwardness in the near-term future.
In the meantime, you should be able to get around this by always explicitly referring to the individual Definition headers, e.g., as "dd/Definitions.hpp".

---

The problem is the other way round, the code should include the `include/Defintions.hpp` (not `include/dd/Definitions.hpp`). I'm not aware of a mechanism to decide which file takes priority from within C++ :/

---

I'm not aware of a way to specify which `Definitions.hpp` should take priority from within C++ and the code that fails is referring to the top-level file (not within `dd` or `zx`) :/

---

Hm. In that case, I do not believe this can be resolved here, but some change has to happen in MQT Core.
I see two paths forward:
 - Remove all variations of https://github.com/cda-tum/mqt-core/blob/3aa78ea7d98f98182011176a07e8a890df266e33/src/CMakeLists.txt#L113-L114 for all derived targets. This would probably require some fixing of includes, but would otherwise resolve the problem in quite an elegant way. For example, everything from the DD package would then be accessible via `#include dd/...`.
 - Modify the duplicate files to give them unique names. Kind of an ugly solution, but would work as well.

---

I feel like the first suggestion would be the cleaner one as you said. I cannot estimate how much effort this will cost :S

---

If I had a guess, a maximum of an hour. It's just deleting some lines from the CMake files and fixing some includes.
I can take care of that if you like.

---

That would be nice 🙂

---

All resolved. Feel free to merge.